Document decisions and results (#55)
Music-Bot-for-Jitsi/Jimmi: docs/ARCHITECTURE.md

[comment]: <> "LTeX: language=en-US"

# Architecture Decisions, Automations and Contributions

This file describes all used frameworks and packages, the reasoning behind them and all other architectural decisions as well as used automations and all contributions made during this project.

---

## Packages and Frameworks

### Deno

> We use [deno.land](https://deno.land), the evolution of Node.js, as runtime for our TypeScript backend. Compared to [Node.js](https://nodejs.org/en/), Deno allows for much greater granularity when it comes to security permissions and has a good *secure-by-default*-ruleset to limit what applications can do if no special permissions given.

Reviewer: ```suggestion
We use [deno.land](https://deno.land), the evolution of Node.js, as runtime for our TypeScript backend. Compared to [Node.js](https://nodejs.org/en/), Deno allows for much greater granularity when it comes to security permissions and has a good *secure-by-default*-rule set to limit what applications can do if no special permissions given.
```

### Svelte

[Svelte](https://svelte.dev/) is a lightweight, modern and fast frontend framework. Svelte has some advantages for which we chose it over other frameworks [source](https://daily.dev/blog/building-with-svelte-all-you-need-to-know-before-you-start):

- Instead of doing most of the work in the browser, Svelte does its work when you build the app, and it compiles it to efficient vanilla JavaScript.
- It reduces the overhead of the JavaScript framework you’re using. Compiling Svelte to vanilla JavaScript improves the code readability, enables re-use, and produces a much faster web app.
- Svelte does not require declarative, state-driven code, which the browser has to convert into DOM operations. That means you don’t need to use virtual DOM anymore.

### Snel

> [Snel](https://deno.land/x/snel@v0.6.0) is a tool/framework to compile Svelte components to JavaScript files using Deno and Svelte to create web applications.

Reviewer 2: ```suggestion
[Snel](https://deno.land/x/snel) is a tool/framework to compile Svelte components to JavaScript files using Deno and Svelte to create web applications.
```
I would prefer to link to the latest version of all tools and frameworks in the documentation

Advantages:

- Simple setup
- Fast compilation
- Hot reloading
- Import maps support
- Support for SCSS and Less out of the box
- TypeScript support

### Opine

[Opine](https://deno.land/x/opine) is a fast, minimalist web framework for Deno ported from [ExpressJS](https://github.com/expressjs/express).
Advantages:

- Robust routing
- Focus on high performance

> - Large selection of HTTP helpers including support for downloading / sending files, etags, Content-Disposition, cookies, JSONP etc.

Reviewer 3: ```suggestion
- Large selection of HTTP helpers including support for downloading/sending files, etags, Content-Disposition, cookies, JSONP etc.
```

- Support for static serving of assets
- View system supporting template engines
- Content negotiation
- Compatible with SuperDeno for easy server testing
- Supports HTTP proxy middleware with opine-http-proxy

### Trex and Velociraptor

[Trex](https://deno.land/x/trex) is a package management tool for Deno similar to NPM but keeping close to the Deno philosophy. Packages are cached and only one `import_map.json` file is generated.  
[Velociraptor](https://deno.land/x/velociraptor) is a script runner for Deno, inspired by NPM's `package.json` scripts. It offers a similar experience but with out-of-the-box support for declarative Deno CLI options, environment variables, concurrency and git hooks.  
In this project we use Velociraptor to have a single command that starts frontend and backend (or other tasks) in parallel.

## Automations

### API Client Generation

> Our backend has a REST-API in which every endpoint has a swagger API specification, written as JSDoc comments over each endpoint function. [This pipeline](https://github.com/Music-Bot-for-Jitsi/Jimmi/blob/main/.github/workflows/update-API-client.yml) is triggered an any backend change and will automatically generate, commit and push a deno compatible API client using the [openAPI code generator](https://github.com/Music-Bot-for-Jitsi/Jimmi/issues/44). The generated API client is saved as a separate deno module in the repository [jimmi-api-client](https://github.com/Music-Bot-for-Jitsi/jimmi-API-client) which is also included as git submodule in [`frontend/lib/jimmi-api-client`](https://github.com/Music-Bot-for-Jitsi/Jimmi/tree/main/frontend/lib) for easier development.

Reviewer 3: ```suggestion
Our backend has a REST-API in which every endpoint has a swagger API specification, written as JSDoc comments over each endpoint function. [This pipeline](https://github.com/Music-Bot-for-Jitsi/Jimmi/blob/main/.github/workflows/update-api-client.yml) is triggered on any backend change and will automatically generate, commit and push a deno compatible API client using the [openAPI code generator](https://github.com/Music-Bot-for-Jitsi/Jimmi/issues/44). The generated API client is saved as a separate deno module in the repository [jimmi-api-client](https://github.com/Music-Bot-for-Jitsi/jimmi-API-client) which is also included as git submodule in [`frontend/lib/jimmi-api-client`](https://github.com/Music-Bot-for-Jitsi/Jimmi/tree/main/frontend/lib) for easier development.
```
